opm/grid/GraphOfGridWrappers.hpp

```diff
@@ -0,0 +1,270 @@
+    bool checkWellIntersections=true)
+{
+    // create compressed lookup from cartesian.
+    const auto& grid = gog.getGrid();
+    const auto& cpgdim = grid.logicalCartesianSize();
+    std::vector<int> cartesian_to_compressed(cpgdim[0]*cpgdim[1]*cpgdim[2], -1);
+    for( int i=0; i < grid.numCells(); ++i )
+    {
+        cartesian_to_compressed[grid.globalCell()[i]] = i;
+    }
+
+    for (const auto& w: wells)
+    {
+        std::set<int> wellsgID;
+        for (const int& cell : w.second)
+        {
+            int gID = cartesian_to_compressed.at(cell);
+            assert(gID!=-1); // well should be an active cell
+            wellsgID.insert(gID);
+        }
+        gog.addWell(wellsgID,checkWellIntersections);
+    }
+}
+
+/// \brief Add well cells' global IDs to the list
+///
+/// Output of the partitioning is missing vertices that were contracted.
+/// This function fills in omitted gIDs and gives them the properties
+/// (like process number and ownership) of their representative cell (well ID).
+template<typename TheTuple>
+void extendImportExportList (const GraphOfGrid<Dune::CpGrid>& gog,
+                                       std::vector<TheTuple>& cellList)
+{
+    // using TheTuple = std::tuple<int,int,char>; or std::tuple<int,int,char,int>
+    using CellList = std::vector<TheTuple>;
+    // make a list of wells for easy identification. Contains ID, begin, end
+    using iter = std::set<int>::const_iterator;
+    std::list<std::tuple<int,iter,iter>> wellList;
+    for (const auto& well : gog.getWells())
+    {
+        wellList.push_front(std::make_tuple(*well.begin(),well.begin(),well.end()));
+    }
+
+    CellList addToList;
+    // iterate once through the original cellList
+    for (const auto& cell : cellList)
+    {
+        // if a cell is in any well, add cells of the well to cellList
+        // and remove the well from the wellList to quicken the search
+        for (auto wID=wellList.begin(); wID!=wellList.end(); )
+        {
+            if (std::get<0>(*wID) == std::get<0>(cell))
+            {
+                for (auto pgID = std::get<1>(*wID); pgID!=std::get<2>(*wID); ++pgID)
+                {
+                    // cells in one well have the same attributes (except ID)
+                    if (*pgID!=std::get<0>(cell))
+                    {
+                        TheTuple wellCell = cell;
+                        std::get<0>(wellCell) = *pgID;
+                        addToList.push_back(wellCell);
+                    }
+                }
+                wID = wellList.erase(wID);
+            }
+            else
+            {
+              ++wID;
+            }
+        }
+        if (wellList.empty())
+        {
+            break;
+        }
+    }
+    std::sort(addToList.begin(),addToList.end(),[](const auto& a, const auto& b){return std::get<0>(a)<std::get<0>(b);});
+    int origSize = cellList.size();
+    int totsize = origSize+addToList.size();
+    cellList.reserve(totsize);
```

Comment:
I think we do know who many entries will be added in the function already at the beginning, right?
In that case we might be able to omit the extra container addToList and push_back to cellList directly.

I might be missing something here, though.

Reply:
The number of cells to be added to the cellList depends on which wells are in the list. Unless all wells are always in the list we do not know the final size beforehand.